File: java/src/main/java/com/brq/payment/StressTest.java
package com.brq.payment;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class StressTest {
    // URL base da API de pagamentos
    private static final String BASE_URL = "http://localhost:8080/api/v1";
    
    // Configuração do cliente HTTP com timeout de 10 segundos
    private static final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    public static void main(String[] args) throws Exception {
        System.out.println("Iniciando teste de stress da API de Pagamentos");
        System.out.println("============================================");
        
        // Configurações dos testes com diferentes cargas
        int[][] testConfigs = {
            {100, 10},  // Teste de aquecimento
            {500, 20},  // Carga moderada
            {1000, 30}  // Carga alta
        };

        for (int i = 0; i < testConfigs.length; i++) {
            runStressTest(testConfigs[i][0], testConfigs[i][1]);
            
            // Adiciona pausa somente se não for o último teste
            if (i < testConfigs.length - 1) {
                System.out.println("\nAguardando 10 segundos antes do próximo teste...");
                Thread.sleep(10000);
            }
        }
    }

    private static void runStressTest(int totalTransactions, int concurrentRequests) throws Exception {
        System.out.printf("\nIniciando teste com %d transações e %d requisições concorrentes\n", 
            totalTransactions, concurrentRequests);
        System.out.println("----------------------------------------------------");

        // Verificar disponibilidade do servidor
        if (!checkServerAvailability()) {
            System.out.println("\n⚠️ ERRO: Servidor não está disponível em " + BASE_URL);
            System.out.println("Por favor, verifique se o servidor está em execução.");
            System.out.println("Comando para iniciar o servidor: mvn spring-boot:run");
            return;
        }

        System.out.println("✅ Servidor disponível, iniciando testes...\n");

        ExecutorService executor = Executors.newFixedThreadPool(concurrentRequests);
        List<CompletableFuture<Boolean>> futures = new ArrayList<>();
        long startTime = System.currentTimeMillis();
        List<Long> responseTimes = new ArrayList<>();
        int successfulRequests = 0;
        int failedRequests = 0;

        // Processamento em lotes para melhor controle
        int batchSize = Math.min(500, totalTransactions);
        for (int i = 0; i < totalTransactions; i += batchSize) {
            int currentBatchSize = Math.min(batchSize, totalTransactions - i);
            System.out.printf("Processando lote %d (%d transações)\n", (i/batchSize + 1), currentBatchSize);

            for (int j = 0; j < currentBatchSize; j++) {
                CompletableFuture<Boolean> future = CompletableFuture.supplyAsync(() -> {
                    try {
                        long requestStart = System.currentTimeMillis();
                        boolean success = sendTransaction();
                        responseTimes.add(System.currentTimeMillis() - requestStart);
                        return success;
                    } catch (Exception e) {
                        System.out.println("Erro na requisição: " + e.getMessage());
                        return false;
                    }
                }, executor);
                futures.add(future);
            }

            // Aguarda conclusão do lote
            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
            
            // Contabiliza resultados
            for (CompletableFuture<Boolean> future : futures) {
                if (future.get()) successfulRequests++;
                else failedRequests++;
            }
            futures.clear();
            
            System.out.println("Lote processado. Pausa de 500ms para controle de carga...");
            Thread.sleep(500);
        }

        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);

        // Cálculo das métricas
        long totalTime = System.currentTimeMillis() - startTime;
        double tps = successfulRequests / (totalTime / 1000.0);
        responseTimes.sort(null);
        double p99Latency = responseTimes.get((int)(responseTimes.size() * 0.99));

        // Impressão dos resultados
        System.out.println("\n=== Resultados do Teste de Stress ===");
        System.out.printf("Tempo total: %.2f segundos\n", totalTime/1000.0);
        System.out.printf("Requisições bem-sucedidas: %d\n", successfulRequests);
        System.out.printf("Requisições com falha: %d\n", failedRequests);
        System.out.printf("Transações por segundo (TPS): %.2f\n", tps);
        System.out.printf("Latência P99: %.2f ms\n", p99Latency);

        // Obtém métricas da API
        printApiMetrics();
    }

    private static boolean sendTransaction() throws Exception {
        String json = String.format("""
            {"accountID":"%s","amount":100.50,"type":"credit"}
            """, UUID.randomUUID());

        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(BASE_URL + "/transactions"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(json))
            .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        return response.statusCode() == 200;
    }

    private static boolean checkServerAvailability() {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/metrics"))
                .GET()
                .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            return response.statusCode() == 200;
        } catch (Exception e) {
            System.out.println("Erro ao verificar disponibilidade: " + e.getMessage());
            return false;
        }
    }

    private static void printApiMetrics() {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/metrics"))
                .GET()
                .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            System.out.println("\n=== Métricas da API ===");
            System.out.println(response.body());
        } catch (Exception e) {
            System.out.println("Falha ao obter métricas da API: " + e.getMessage());
        }
    }
}

File: stress_test.py
import asyncio
import aiohttp
import time
import uuid
import os

async def send_transaction(session, transaction_data, timeout=10):
    try:
        async with session.post(
            'http://localhost:8000/api/v1/transactions', 
            json=transaction_data,
            timeout=timeout
        ) as response:
            return await response.json()
    except (aiohttp.ClientConnectorError, asyncio.TimeoutError, 
            aiohttp.ServerDisconnectedError) as e:
        print(f"Erro na requisição: {type(e).__name__}")
        await asyncio.sleep(0.5)
        return None

async def check_server_availability():
    try:
        timeout = aiohttp.ClientTimeout(total=5)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get('http://localhost:8000/api/v1/metrics') as response:
                return response.status == 200
    except Exception as e:
        print(f"Erro ao verificar disponibilidade do servidor: {e}")
        return False

async def run_stress_test(total_transactions=1000, concurrent_requests=50):
    print(f"\nIniciando teste de stress com {total_transactions} transações...")
    print(f"Requisições concorrentes: {concurrent_requests}")
    
    # Verificar se o servidor está disponível
    server_available = await check_server_availability()
    if not server_available:
        print("\n⚠️ ERRO: Servidor não está disponível em http://localhost:8000")
        print("Por favor, verifique se o servidor está rodando com o comando:")
        print("python main.py")
        return
    
    # Configurações otimizadas para o cliente HTTP - com limites mais razoáveis
    conn = aiohttp.TCPConnector(limit=concurrent_requests, ttl_dns_cache=300, force_close=True)
    timeout = aiohttp.ClientTimeout(total=30, connect=10, sock_connect=10, sock_read=10)
    
    successful_requests = 0
    failed_requests = 0
    
    async with aiohttp.ClientSession(connector=conn, timeout=timeout) as session:
        start_time = time.time()
        response_times = []

        # Dividir as transações em lotes menores para evitar sobrecarga
        batch_size = min(500, total_transactions)
        
        for batch_start in range(0, total_transactions, batch_size):
            batch_end = min(batch_start + batch_size, total_transactions)
            print(f"Processando lote {batch_start//batch_size + 1} ({batch_end-batch_start} transações)")
            
            # Processar em mini-lotes para melhor controle
            mini_batch_size = min(concurrent_requests, 50)
            
            for mini_batch_start in range(batch_start, batch_end, mini_batch_size):
                mini_batch_end = min(mini_batch_start + mini_batch_size, batch_end)
                
                # Criar tarefas para o mini-lote
                tasks = []
                for _ in range(mini_batch_start, mini_batch_end):
                    transaction_data = {
                        "accountID": str(uuid.uuid4()),
                        "amount": 100.50,
                        "type": "credit"
                    }
                    
                    request_start = time.time()
                    task = asyncio.create_task(send_transaction(session, transaction_data))
                    tasks.append((task, request_start))
                
                # Aguardar todas as tarefas do mini-lote
                for task, req_start in tasks:
                    try:
                        result = await task
                        if result is not None:
                            successful_requests += 1
                            response_times.append(time.time() - req_start)
                        else:
                            failed_requests += 1
                    except Exception as e:
                        failed_requests += 1
                        print(f"Erro ao processar resultado: {e}")
                
                # Pausa entre mini-lotes para evitar sobrecarga
                await asyncio.sleep(0.5)
            
            # Pausa maior entre lotes principais
            print(f"Lote {batch_start//batch_size + 1} concluído. Aguardando...")
            await asyncio.sleep(2)

        total_time = time.time() - start_time
        tps = successful_requests / total_time if total_time > 0 else 0
        
        # Calcula P99 latência
        if response_times:
            sorted_times = sorted(response_times)
            p99_index = int(len(sorted_times) * 0.99)
            p99_latency = sorted_times[p99_index] * 1000 if p99_index < len(sorted_times) else 0
        else:
            p99_latency = 0

        print("\n=== Resultados do Teste de Stress ===")
        print(f"Tempo total: {total_time:.2f} segundos")
        print(f"Requisições bem-sucedidas: {successful_requests}")
        print(f"Requisições com falha: {failed_requests}")
        print(f"Transações por segundo (TPS): {tps:.2f}")
        print(f"Latência P99: {p99_latency:.2f}ms")
        
        # Verificar métricas da API
        try:
            async with session.get('http://localhost:8000/api/v1/metrics') as response:
                api_metrics = await response.json()
                print("\n=== Métricas da API ===")
                print(f"TPS registrado pela API: {api_metrics['tps']}")
                print(f"P99 Latência pela API: {api_metrics['p99_latency']}ms")
                print(f"Total de requisições: {api_metrics['total_requests']}")
        except Exception as e:
            print(f"Erro ao obter métricas: {e}")

if __name__ == "__main__":
    print("Iniciando teste de performance da API de pagamentos")
    
    # Configurações de CPU para melhor desempenho
    cpu_count = os.cpu_count()
    print(f"Número de CPUs disponíveis: {cpu_count}")
    
    # Configurações de teste otimizadas para latência
    test_configs = [
        (100, 10),     # Teste de aquecimento
        (500, 20),     # Teste com carga moderada
        (1000, 30)     # Teste com carga mais alta
        
    ]
    
    for i, (transactions, concurrency) in enumerate(test_configs):
        asyncio.run(run_stress_test(transactions, concurrency))
        if i < len(test_configs) - 1:
            print("\nAguardando 10 segundos antes do próximo teste...")
            time.sleep(10)
